dir.c

```diff
@@ -1357,25 +1360,53 @@ static enum path_treatment treat_directory(struct dir_struct *dir,
 		return path_none;
 
 	case index_nonexistent:
-		if (dir->flags & DIR_SHOW_OTHER_DIRECTORIES)
-			break;
+		if (dir->flags & DIR_SHOW_OTHER_DIRECTORIES) {
+			if (dir->flags & DIR_HIDE_EMPTY_DIRECTORIES) {
+				break;
+			}
+			else {
+				return exclude ? path_excluded : path_untracked;
+			}
+		}
+		if (exclude &&
+			(dir->flags & DIR_SHOW_IGNORED_TOO) &&
+			(dir->flags & DIR_COLLAPSE_IGNORED)) {
+				stop_at_first_file = 1;
+				break;
+		}
 		if (!(dir->flags & DIR_NO_GITLINKS)) {
 			unsigned char sha1[20];
 			if (resolve_gitlink_ref(dirname, "HEAD", sha1) == 0)
 				return exclude ? path_excluded : path_untracked;
 		}
+
 		return path_recurse;
 	}
 
-	/* This is the "show_other_directories" case */
-
-	if (!(dir->flags & DIR_HIDE_EMPTY_DIRECTORIES))
-		return exclude ? path_excluded : path_untracked;
-
+	/*
+	 * We need to recurse to check the path_treatment for this
+	 * directory.
+	 */
 	untracked = lookup_untracked(dir->untracked, untracked,
 				     dirname + baselen, len - baselen);
-	return read_directory_recursive(dir, dirname, len,
-					untracked, 1, pathspec);
+
+	read_directory_recursive_result = read_directory_recursive(dir, dirname, len,
+					untracked, 1, stop_at_first_file, pathspec);
+
+	/*
+	 * If stop_at_first_file is set, this means that the path (directory)
+	 * matches an exclude pattern. We only need to recurse to
+	 * check if the directory is empty or not. If directory
+	 * is not empty, we know it is a non-empty excluded directory.
+	 */
+	if (stop_at_first_file) {
```

Comment:
I believe you are correct here.  You would still need the `stop_at_first_file` variable because it is passed into the call to `read_directory_recursive()` above.  I'd lean towards keeping this as is for now, and considering this change for a second PR because the changes are independent.